kythe/java/com/google/devtools/kythe/platform/kzip/KZipReader.java

```diff
@@ -0,0 +1,109 @@
+/*
+ * Copyright 2018 Google Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.google.devtools.kythe.platform.kzip;
+
+import com.google.common.io.ByteStreams;
+import com.google.devtools.kythe.proto.Analysis;
+import com.google.gson.Gson;
+import com.google.gson.GsonBuilder;
+import java.io.File;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+import java.util.Enumeration;
+import java.util.Iterator;
+import java.util.zip.ZipEntry;
+import java.util.zip.ZipFile;
+
+/** Read a kzip file. */
+public final class KZipReader implements KZip.Reader {
+
+  private final Gson gson;
+  private final ZipFile zipFile;
+  private final String rootPrefix;
+
+  public KZipReader(File file) throws IOException {
+    this(file, KZip.buildGson(new GsonBuilder()));
+  }
+
+  public KZipReader(File file, GsonBuilder gsonBuilder) throws IOException {
+    this(file, KZip.buildGson(gsonBuilder));
+  }
+
+  public KZipReader(File file, Gson gson) throws IOException {
+    this.gson = gson;
+    this.zipFile = new ZipFile(file, ZipFile.OPEN_READ);
+    this.rootPrefix = getRootPrefix(this.zipFile);
+  }
+
+  private static String getRootPrefix(ZipFile zipFile) {
+    Enumeration<? extends ZipEntry> zipEntires = zipFile.entries();
+    if (!zipEntires.hasMoreElements()) {
+      throw new KZipException("missing root entry");
+    }
+    ZipEntry root = zipEntires.nextElement();
+    if (!root.isDirectory()) {
+      throw new KZipException("invalid root entry: " + root.getName());
+    }
+    String rootPrefix = root.getName();
+
+    // Make sure each path in the kzip has the same root.
+    while (zipEntires.hasMoreElements()) {
+      ZipEntry zipEntry = zipEntires.nextElement();
+      if (!zipEntry.getName().startsWith(rootPrefix)) {
+        throw new KZipException("Invalid entry (bad root): " + zipEntry.getName());
+      }
+    }
+    return rootPrefix;
+  }
+
+  @Override
+  public Iterator<Analysis.IndexedCompilation> scan() {
```

Comment:
It can, but that would mean collapsing the stream unless there is a method I don't know about. I agree that the syntax with an iterable is much nicer for the user, so I could go either way.

Reply:
I believe you can just return `() -> zipFile.stream()...iterator()` as an `Iterable<Analysis.IndexedCompilation>`

Reply:
yup, thanks!